### PR TITLE
Remove type links and use global vars

### LINKED ISSUE #1437: Remove typelinks (`tl`) in favor of  the global variables
Since the introduction #1407 of the global variables, the typelink concept becomes reduntant. We shoulld:
- Remove the `tl` commands
- Remove the `librz/analysis/typelink.c`
- Port the analysis code from using `rz_analysis_type_link_*` API to use types of the global variables instead
- Port the disassembly output to use global variables instead of the typelinks
- Remove the typelinks from the project format and implement conversion from typelinks to global vars in `librz/core/project_migrate.c`

```
[i] ℤ rg rz_analysis_type_link                                                                                                                                                                                                    12:53:55 
librz/include/rz_analysis.h
1624:RZ_API bool rz_analysis_type_link_exists(RzAnalysis *analysis, ut64 addr);
1625:RZ_API RZ_BORROW RzType *rz_analysis_type_link_at(RzAnalysis *analysis, ut64 addr);
1629:RZ_API bool rz_analysis_type_link_offset(RzAnalysis *analysis, RZ_BORROW RzType *type, ut64 addr);
1630:RZ_API RZ_OWN RzList /* RzType */ *rz_analysis_type_links(RzAnalysis *analysis);
1631:RZ_API RZ_OWN RzList /* RzTypePath */ *rz_analysis_type_links_by_offset(RzAnalysis *analysis, ut64 offset);

librz/core/ctypes.c
983:    RzType *link = rz_analysis_type_link_at(core->analysis, addr);

librz/core/disasm.c
5326:           RzType *link_type = rz_analysis_type_link_at(core->analysis, ds->addr + idx);

librz/analysis/typelink.c
16:RZ_API bool rz_analysis_type_link_exists(RzAnalysis *analysis, ut64 addr) {
31:RZ_API RZ_BORROW RzType *rz_analysis_type_link_at(RzAnalysis *analysis, ut64 addr) {
96:RZ_API RZ_OWN RzList /* RzType */ *rz_analysis_type_links(RzAnalysis *analysis) {
142:RZ_API RZ_OWN RzList /* RzTypePath */ *rz_analysis_type_links_by_offset(RzAnalysis *analysis, ut64 offset) {
```
These typelinks are used in `librz/core/ctypes.c` in `resolve_type_links()` and subsequently in `rz_core_link_stroff()` functions. which is then used in the analysis for the `aat` command:
```
[i] ℤ rg rz_core_link_stroff                                                                                                                                                                                                      13:03:12 
librz/include/rz_core.h
526:RZ_API void rz_core_link_stroff(RzCore *core, RzAnalysisFunction *fcn);

librz/core/ctypes.c
761:RZ_API void rz_core_link_stroff(RzCore *core, RzAnalysisFunction *fcn) {
977:            rz_core_link_stroff(core, fcn);

librz/core/cmd_analysis.c
7936:                           rz_core_link_stroff(core, fcn);
7949:                           rz_core_link_stroff(core, fcn);
```

See also `global` variables loading in `librz/analysis/dwarf_processing.c`

_Originally posted by @XVilka in https://github.com/rizinorg/rizin/issues/1407#issuecomment-896502594_

### LINKED ISSUE #1495: Global variables should be replaced in disassembly
**Is your feature request related to a problem? Please describe.**
Even if I have defined a global variable with `avg`, this does not improve the readability of the disassembly.

**Describe the solution you'd like**
- [ ] global variable name used when the address is used in the disassembly
- [ ] if the variable is a struct or a complex type, show also accesses to inner fields

**Describe alternatives you've considered**
.

**Additional context**
.

Reference: https://github.com/rizinorg/rizin/issues/333

Depends on https://github.com/rizinorg/rizin/issues/1437

### PR DESCRIPTION
**Pull request by XVilka**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [x] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

- Remove type links from RzAnalysis
- Remove `tl` commands
- Switch to using RzAnalysisVarGlobal (`avg` commands everywhere)
- Updated Project version and added migrations and conversions
- Changed `avga` (add global var) syntax from `avga name offset type` to `avga name type` and using `@ 0xXXXX` for offset setting
- Now type offset propagation from global variables requires running `aat` (before it was invoked as a side effect of creating a typelink, with the current offset where command is running as the reference)

**Test plan**

CI is green

**Closing issues**

- Closes https://github.com/rizinorg/rizin/issues/1437
- Closes https://github.com/rizinorg/rizin/issues/1495

### CODE REVIEW

**Review by wargio — APPROVED** [review]

File: librz/core/project_migrate.c [hunk 1]
```diff
@@ -500,6 +475,62 @@ RZ_API bool rz_project_migrate_v11_v12(RzProject *prj, RzSerializeResultInfo *re
...
+// Migration 12 -> 13
+//
+// Changes from 5badfe10f655374f424164d62915cb3a1d22e2b9:
+//	Removed typelinks from "/core/analysis/typelinks"
+//	and converted into global variables in "/core/analysis/vars"
+
+typedef struct {
+	RzList /*<char *>*/ *moved_keys; ///< deferred for deletion from the old sdb
+	Sdb *global_vars_db;
+} V3V4TypesCtx;
```
Comment (thestr4ng3r) [inline librz/core/project_migrate.c]:
```suggestion
} V12V13TypesCtx;
```

and other namings below

File: librz/analysis/types.c [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+#include <rz_analysis.h>
+#include <string.h>
+
+static RZ_OWN RzList /* <RzTypePathTuple*> */ *var_global_type_paths(RzAnalysis *analysis, RzAnalysisVarGlobal *gv, ut64 addr) {
+	rz_return_val_if_fail(gv, false);
+
+	RzType *t = gv->type;
+
+	// Handle only identifiers here
+	if (t->kind != RZ_TYPE_KIND_IDENTIFIER) {
```
Comment (thestr4ng3r) [inline librz/analysis/types.c]: Only as a hint, there is also `rz_type_path_by_offset()` which would let you skip a lot of this inlined logic and supports even more kinds of types like arrays. But I see this code was only moved here as-is so changes like that should go in another commit then.

Comment (XVilka) [reply librz/analysis/types.c]: Yes, my plan is to complete migration first, then there are a few aspects of what could be improved: 1) display 2) propagation 3) integration with DWARF/PDB/whatever. The plan is to do it after this PR is merged though, but before 0.6.0

Comment (thestr4ng3r) [reply librz/analysis/types.c]: Agreed this is the way to go.

Comment (XVilka) [reply librz/analysis/types.c]: While solving side effect cancer, I did this anyway, to simplify the code.

**Comment by XVilka** [comment]
Now type offset propagation from global variables requires running `aat` (before it was invoked as a side effect of creating a typelink, with the current offset where command is running as the reference)

**Comment by XVilka** [comment]
Only one test left to fix:
```
[XX] db/cmd/types jni and SoP (preprocessed)
RZ_NOPLUGINS=1 /private/var/folders/bc/yvvw2g2s4jlg9r8x7tf61xv800009d/T/woodpecker-local-1287038090/rizinorg/rizin/prefix/bin/rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -eflirt.sigdb.load.system=false -eflirt.sigdb.load.home=false -N -Qc 'e asm.bytes=true
e asm.lines.bb=false
aa
s sym.Java_com_app_ndh_NDHActivity_print
to bins/headers/jni_preprocessed.h

# TODO: Tracked in https://github.com/rizinorg/rizin/issues/1132
# This is actually quite wrong, regarding what this function actually does :-)
# We are manually crafting the canary value such that it looks like a pointer
# and say that at this pointer there is a JNINativeInterface object, in order
# to test propagation from a global variable.
# Do not hesitate to edit this test if you find a better case for testing this.
e io.cache=1
wx 37133713 @ reloc.target.__stack_chk_guard
avga bla JNINativeInterface @ 0x13371337
avga foo JNINativeInterface @ 0x464c457f # \x7fELF, the "pointer" when dereferencing addr 0
aat

pdf~JNINativeInterface
pdr~JNI?
' bins/elf/libverifyPass.so
-- stdout
--- expected
+++ actual
@@ -1,6 +1,3 @@
-| ; var struct JNINativeInterface JNINativeInterface @ stack - 0x14
-| 0x00000f74      4b93           str   r3, [JNINativeInterface.reserved0]
-| 0x00000f76      039b           ldr   r3, [var_134h]                  ; JNINativeInterface.reserved0
 | 0x00000f7e      d358           ldr   r3, [r2, r3]                    ; JNINativeInterface.GetStringUTFChars
 | 0x00000f82      019a           ldr   r2, [var_13ch]                  ; JNINativeInterface.reserved0
 | 0x0000100e      d358           ldr   r3, [r2, r3]                    ; JNINativeInterface.GetStringLength
@@ -8,6 +5,4 @@
 | 0x00001066      d258           ldr   r2, [r2, r3]                    ; JNINativeInterface.NewStringUTF
 | 0x00001114      d258           ldr   r2, [r2, r3]                    ; JNINativeInterface.NewStringUTF
 | 0x0000112e      d258           ldr   r2, [r2, r3]                    ; JNINativeInterface.NewStringUTF
-| 0x00001144      4b9a           ldr   r2, [JNINativeInterface.reserved0]
-| 0x00001148      9a42           cmp   r2, r3                          ; JNINativeInterface.reserved0
-12
+7

-- stderr
[x] Analyze all flags starting with sym. and entry0 (aa)
ERROR: Typedef "va_list" already exists
Error forming RzType and RzBaseType pair out of typedef: "va_list"
```

**Review by wargio — APPROVED** [review]

**Comment by wargio** [comment]
i would set this as broken

**Comment by XVilka** [comment]
@wargio good idea, I created an issue to track it: https://github.com/rizinorg/rizin/issues/3624
